PyISY Beta Changes (Major Refactoring, Move to PEP8 & Python 3, Climate Fixes, Z-Wave Support, Retry Fixes)

shbatm opened the pull request:
**EDITED 1-JUL-2019**:
I've updated this PR to include a significant refactoring and cleanup of the original PyISY code, with the primary goal of (1) fixing as many bugs in one shot as possible and (2) moving towards PEP8 compliant code with as few breaking changes as possible.

**This fixes:**
- #22, #31, #32, #41, #43, #45, #46
- Malformed climate control commands
   - They were missing the `self._id` parameter, were missing a `.conn` in the command path and did not convert the values to strings before attempting to encode.
   - They are sending *2 for the temperature for ALL thermostats instead of just Insteon/UOM 101.
   - Several modes were missing for the Insteon Thermostats.
- Adds increased Z-Wave support by storing the `devtype` category (since `type` is useless for Z-Wave)
- Adding retries for failed REST calls to the ISY #46
- Fix Node.aux_properties inconsistent typing #43 and now updates the existing aux_props instead of re-writing the entire dict.
- Zwave multisensor support #31 -- Partial Fix, this is being mostly address in my Home Assistant updates. [Forum Thread is here](https://community.home-assistant.io/t/isy994-z-wave-sensor-enhancements-testers-wanted/124188)
- Major code refactoring to consolidate nested function calls, remove redundant code.
- Linting to PEP8 in as many places as possible except where it would cause obvious breaking changes with other integrations.
- Dynamically adds control functions to the Node and Program class--this is for future expansion to only add the appropriate commands to a given Node Type (e.g. don't add climate_mode to a light).

**Known Breaking Changes:**
- A node Event is now returned as an object. In most cases this is a benefit because it returns more details than just the received command (value, uom, precision, etc), and it represents itself in string from the same as previously; however, direct comparisons will now fail unless updated:
    - "`event == "DON"`" must be replaced with "`event.event == "DON"`" or "`str(event) == "DON"`"
- Node Unit of Measure is returned as a string if it is not a list of UOMs, otherwise it is returned as a list. Previously this was returned as a 1-item list if there was only 1 UOM.
    - ISYv4 and before returned the UOM as a string ('%/on/off' or 'degrees'), ISYv5 phases this out and uses numerical UOMs that correspond to a defined value in the SDK (included in constants file).
    - Previous implementations of `unit = uom[0]` should be replaced with `unit = uom` and for compatibility, UOM should be checked if it is a list with `isinstance(uom, list)`.
    ```python
        uom = self._node.uom
        if isinstance(uom, list):
            uom = uom[0]
    ```
- Node.lock() and Node.unlock() methods are now Node.secure_lock() and Node.secure_unlock().
- Most functions *that should be internal only* have been renamed to snake_case from CamelCase.
- Node Parent property has been renamed. Internal property is `node._parent_nid`, but externally accessible property is `node.parent_node`.

**Still in progress:** (everything above works, feel free to test using PyISY-beta PyPi package)

- [x] Program and Variable changes request the full update from ISY for all Programs/Variables. This will be changed to listen for changes only if `auto_update = True`.
- [ ] Add back support for Network Resources to be called directly.

shbatm commented:
Just checking in, have you had a chance to take a look at these changes yet?

shbatm commented:
@rmkraus Just wondering if you've had a chance to consider this change? I'm at a holding point with home-assistant/home-assistant#20970 until this progresses or I have to fork the NPM module.

I know you're busy, just wanted to nudge. Let me know if you have any questions.

shbatm commented:
@OverloadUT -- I just realized I've been pushing changes to a branch that this PR wasn't tracking.  I've merged everything that's been working from the `PyISY-beta` implementation into this branch and set a forecast version of `1.2.0` for the merge (PyISY-beta is on `1.1.7` at the moment and I'll keep bumping it in the `1.1.x` range for now during my testing.

Just letting you know that I've still been actively working on squashing bugs in this branch and didn't want to conflict with anything you change.  Please feel free to reach out with any review comments once you do get a chance to look at this.

shbatm commented:
@overloadut, done pushing changes unless bugs are found. Please let me know if you find any bugs or want to see anything different. Also, curious what you find with the readline issue from the HA forums... I'd love to test the fix for it.

OverloadUT commented:
Wow, this is a doozie. It's going to take a bit to review the whole thing. I've started looking it over and I'm pulling it down to set up in my beta environment

shbatm commented:
Yeah, sorry it's all in one. As I tried to update things it was tough trying to trace through the 4-deep function calls to get back to the ISY. Once I started trying to simplify and lint it just snowballed, but I hope the end goal is a little bit cleaner read and slightly more efficient code. 

Take your time and please test thoroughly. There are several others testing it over on the HA forums. If you have any specific questions let me know.

OverloadUT commented:
Hey shbatm, I see a few more commits here, looks like you're including the fix for #51?

Let me know when it's all stable for formal testing. I saw you mention that you were thinking of adding chunked encoding support to this PR: I would ask that you please don't do that, and let's keep this PR to showstopper bug fixes only now. It's one of those PRs that could never end if we let feature creep happen :)

shbatm commented:
Should be good to test now. Lumped in the suggested change for #51 and squashed a bug found in the HA forums. I'll quit messing around in this branch unless it's a major bug.

For convenience, this is also on PyPi under PyISY_beta.

rmkraus commented on setup.py:
```diff
@@ -1,11 +1,17 @@
 setup(
     name='PyISY',
-    version='1.1.1',
+    version='1.2.0',
```
This should probably be a version 2.y.z. There aren't really changes to the external API, but it is such a big refactoring that it would probably be good to track the change.

OverloadUT replied:
Agreed. I had planned on bumping the major version before publishing this update with a release page writeup of why.

rmkraus replied:
👍

shbatm replied:
👍

shbatm replied:
A bigger question to this--if this is going to be a major version bump, should I go ahead and finish updating the function names/properties to be PEP8 compliant? I left several "externally called" functions the same to not be a breaking change (e.g. `sendX10`), but if we document the changes it wouldn't be much more to fix the rest of the functions.  My original intent was to not introduce any major breaking changes--but there are a few.

OverloadUT replied:
Good question. Let's keep the method names the same with the goal of this update having no "known" big breaking changes. This will increase the number of people that will update to this version and give it some more testing. If we change it now, we'll get fewer eyes on the version from people who don't want to update their code.

Then there's nothing wrong with changing the method names in another major version bump for the next release.

shbatm replied:
Ok. I'm trying to keep track of the breaking changes in the PR description above and will move them into a changelog soon. 

The UOM change may need to be reverted back to the previous 1 item list, there are several places this had to be patched in the HA code which makes everything cleaner overall and preps for ISYv5, but could be a PITA for some other integrations to update. Either way we need to document better what UOM returns for what version of ISY firmware.

rmkraus commented on PyISY/isy.py:
```diff
@@ -0,0 +1,248 @@
+        NOTE: `on` and climate setpoints are special and are still defined
+        explicitly in the Node class.
+        """
+        commands = ['DFON', 'DFOF', 'BRT', 'DIM', 'BEEP',
```
The list should only contain one type. That way you don't have to build logic later to handle the fact that you don't know what you're getting out of the list.

```python
command = [('DFON', None), ..., ('SECMD', '84'), ...]
...
for command, values in commands:
```

Trims down a few lines.

shbatm replied:
Updated. Curious what your thoughts are on this in general: I added it to dynamically add the control functions because they were otherwise causing a lot of redundant code.  Eventually, I could see this being used to only add relevant functions depending on Node types (e.g. don't add climate setpoint functions to a light switch)--for now it just eliminates redundant code.

rmkraus replied:
I guess I have three questions on that:
1) How much work is it to make this dependent on the node type?
2) Will it break backwards compatibility with ISY firmware 4?
3) Do we care about question 2?

shbatm replied:
1. More than I care to put forth for this update but doable down the road if it was a priority. The logic is built for evaluating types, but it's in Home Assistant's implementation, not this library.
2. Compatibility would be dependent on how you evaluate the types, HA still relies pretty heavily on insteon types as the primary method for sorting, which works for 4 and 5. My changes have expanded sorting for zwave which is also the same, but v5 node servers is a different story.
3. Until v5 has a timeline for exiting beta, besides Polyisy in October, I think v4 still needs to be supported. I've had at least a few people on the HA forum still using v4.

rmkraus replied:
So this will work with 4, 5, but probably not node servers? If so, I’m ok with that.

rmkraus commented on PyISY/events.py:
```diff
@@ -0,0 +1,299 @@
+            except socket.error:
+                loop = False
+            else:
+                siz = len(new_data)
```
Same as my comment on #52, no need for this variable to be defined.

rmkraus commented on PyISY/events.py:
```diff
@@ -0,0 +1,299 @@
+                            self._routemsg(data)
+
+
+class EventEmitter:
```
I don't remember where these classes were before, but they probably shouldn't be in the same file as the event stream code. I don't believe they are related.

shbatm replied:
Moved back to the `Nodes/__init__.py` file.  I was trying to reorganize, but you're right, these are only used for nodes.

shbatm replied:
Moved theses into a separate `handlers.py` file to prevent circular import errors (could not import EventEmitter when isy.py imports nodes).

rmkraus commented on PyISY/connection.py:
```diff
@@ -0,0 +1,256 @@
+    def __init__(self, tls_ver):
+        """Initialize the TLSHttpAdapter class."""
+        if tls_ver == 1.1:
+            self.tls = getattr(ssl, 'PROTOCOL_TLSv1_1')
```
self.tls = ssl.PROTOCOL_TLSv1_1

rmkraus commented on PyISY/connection.py:
```diff
@@ -0,0 +1,256 @@
+        if tls_ver == 1.1:
+            self.tls = getattr(ssl, 'PROTOCOL_TLSv1_1')
+        elif tls_ver == 1.2:
+            self.tls = getattr(ssl, 'PROTOCOL_TLSv1_2')
```
self.tls = ssl.PROTOCOL_TLSv1_2

rmkraus commented on PyISY/configuration.py:
```diff
@@ -1,10 +1,13 @@
+"""ISY Configuration Lookup."""
 from xml.dom import minidom
```
Is this still needed?

shbatm replied:
Is `minidom` still needed, or the whole configuration file? `minidom` is still used in the `parse` function of this file.  As far as I can tell, the `configuration` lookup is used to check if the climate and networking modules are enabled.

rmkraus replied:
👍

rmkraus commented:
I marked up a couple of things. I like the intent of these changes. They are big, however, so testing is key to making sure nothing breaks.

shbatm commented:
Added a fix for a casting error in Groups where it was trying to do a greater than comparison between a string and a integer.  Also handled unknown states in Groups (besides just status = None).

shbatm commented:
Pushed a bugfix from here: https://community.home-assistant.io/t/isy994-z-wave-sensor-enhancements-testers-wanted/124188/79

Separate subject: what are your thoughts on Black for formatting?  Other than I'm rather pissed I wasted all that time reformatting this code mostly by hand, it seems like a great tool.  If you're not opposed to it I was going to run it over the code I've already changed, but I don't want to mess up @OverloadUT  on your review.  Should be no functional changes, only formatting; and I've changed so much formatting already I don't know that you'll see much difference in the diff's.

OverloadUT commented:
I'm happy with Black, especially now that Hass is using it officially. But let's put that in a separate PR after this one is merged as there's not much value in putting it in to this one.

shbatm commented:
Pushed bugfix from here: https://community.home-assistant.io/t/isy994-z-wave-sensor-enhancements-testers-wanted/124188/86

shbatm commented:
Is there anything I can do to help make this review easier?  I know it's a beast, but let me know if I can help in any way. I could even try to send several smaller PRs to a different (`v2`?) branch.

OverloadUT commented:
Hi @shbatm - I've begun reviewing this and hope to put in some real time on it this weekend!

OverloadUT commented:
In the meantime, can you rebase to resolve the merge conflicts?

shbatm commented:
@OverloadUT Rebase completed. Sorry it took so long, got buried in my to-dos.

OverloadUT commented on README.md:
```diff
@@ -0,0 +1,28 @@
+
+WEBSITE: http://automic.us/portfolio/pyisy
+
+GITHUB: https://github.com/automicus/PyISY
```
This needs to change to be appropriate for being in the main PyISY repo

shbatm replied:
Fixed in latest commit.

OverloadUT commented on setup.py:
```diff
@@ -1,4 +1,6 @@
-from setuptools import setup, find_packages
+import pathlib
```
Unused import

shbatm replied:
Merging error. Updated the setup file to use markdown instead of text. Black Formatted and Linted the setup.py file in the latest commit.

shbatm commented on PyISY/climate.py:
```diff
@@ -1,290 +1,275 @@
-                setattr(self, cname + '_units', unit)
-
-                self.parent.log.info('ISY Updated Climate Value: ' + cname)
+"""ISY Climate Module Information."""
```
I'm for scrapping it, but I would say we wait until the official retirement on 30-Mar-2020 and we do that in a separate PR.

shbatm commented on PyISY/connection.py:
```diff
@@ -0,0 +1,256 @@
+            self._tls_ver = tls_ver
+            # Most SSL certs will not be valid. Let's not warn about them.
+            requests.packages.urllib3.disable_warnings()
+
```
This is [original](https://github.com/automicus/PyISY/blob/58b477038d0be5ff0c3338265827cc81eca0674c/PyISY/Connection.py#L30). Added to #61 to make sure the warnings are handled properly. I vaguely remember taking this out and my connection not working with SSL/TLS.

shbatm commented on PyISY/connection.py:
```diff
@@ -0,0 +1,256 @@
+                                   TLSHttpAdapter(tls_ver))
+        else:
+            self.use_https = False
+            self._tls_ver = None
```
Agree. Let's fix this now. I'll update to include a fix in this branch.

shbatm replied:
@OverloadUT Make sure you're OK with how I changed this. I split the `use_https` and test into two if statements.

shbatm commented:
Thanks for reviewing. I'll address the few issues in the next few days when I get a chance.

I know we had some conversation a while back about if this should be a major version change or not. Since I've had to make additional changes for the beta version, I'm more inclined to make this a major version, bail on Python 2.7 support (since it's no longer supported). What do you think?

I would propose: 
1) I'll make the changes from the review above. 
2) Merge this PR to a new temporary v2 branch
3) I'll send some smaller PRs with the rest of the changes made in my beta branch.
4) Release as a new major version with the known breaking changes documented. Non-Home-Assistant users can keep their version pinned at the current version until the changes are addressed.

shbatm commented on PyISY/Nodes/group.py:
```diff
@@ -1,66 +1,54 @@
+        elif int(self.status) <= 0:
             clean_status = 0
-        if self.status != clean_status:
+        if int(self.status) != clean_status:
```
Fixed in latest commit. Note: these casting issues are one of the things that would be fixed if we move away from VarEvents.

shbatm commented on PyISY/Programs/__init__.py:
```diff
@@ -55,25 +62,25 @@ def __init__(self, parent, root=None, pids=None, pnames=None,
+        ind = self.pids.index(self.root)
+        if self.ptypes[ind] == ATTR_FOLDER:
+            return 'Folder ({})'.format(self.root)
+        if self.ptypes[ind] == 'program':
```
I think I ran out of steam when I got to programs. I didn't touch as much on that side. Updated `'program'` to `ATTR_PROGRAM` constant in the latest commit. There are still some more in that folder that need to be cleaned up, but I'll hold that for a subsequent PR.

shbatm commented on PyISY/Programs/folder.py:
```diff
@@ -17,85 +19,52 @@ class Folder(object):
+                                             command])
+        result = self.isy.conn.request(req_url)
+        if not result:
+            self.isy.log.warning('ISY could not run program: %s', self._id)
```
Fixed in latest commit.

shbatm commented on PyISY/connection.py:
```diff
@@ -0,0 +1,256 @@
+        self.isy.log.error('Bad ISY Request: %s %s: '
+                           'Failed after %s retries',
+                           url, req.status_code, retries)
+        return None
```
Fixed in latest commit.

OverloadUT commented:
@shbatm That plan sounds perfect to me!

shbatm commented:
@OverloadUT -- just pushed the commit referenced in all my comments above. I'll branch off the next PR from here and target whatever you call the new branch.

OverloadUT commented:
Hm, looks like I can't switch this PR to merge to the v2 branch I just created. Could you open a new one against that branch?

shbatm commented:
#62 Targets the new branch.
